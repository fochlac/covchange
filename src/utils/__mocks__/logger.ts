export default jest.fn(console.log)
