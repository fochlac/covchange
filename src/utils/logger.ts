import { createWriteStream, existsSync, mkdirSync } from 'fs'

/**
 *   Log Levels:
 *   0: no error, very important info
 *   1: error: critical error in app
 *   2: error: internal error in app
 *   3: error: minor error
 *
 *   4: info: important info (bad request/login)
 *   5: info: minor info
 *
 *   6: trace: tracing informations
 *   7: unimportant tracing information
 *
 *  10: full trace: (requests objects, etc)
 *
 **/
let currentDate = getDate().day
if (!existsSync(global.appRoot + 'log')) {
	mkdirSync(global.appRoot + 'log')
}

let logStream = createWriteStream(global.appRoot + 'log/output' + getDate().day + '.txt', {
	flags: 'a',
	encoding: 'utf8',
	autoClose: true,
})

logStream.on('error', err => {
	console.log(err)
})

export default (level: number, ...message: any[]): void => {
	const now = getDate()
	const messages = message.map(parseMessage).join(' ')
	const logMessage = `${now.day} - ${now.time} - ${level} - ${messages}`

	if (level <= global.logLevel) {
		console.log(logMessage)
	}

	if (currentDate !== getDate().day) {
		currentDate = getDate().day
		logStream.end()
		logStream = createWriteStream(global.appRoot + 'log/output' + getDate().day + '.txt', {
			flags: 'a',
			encoding: 'utf8',
			autoClose: true,
		})
	}

	logStream.write(logMessage)
}

function parseMessage(item) {
	let output
	if (['string', 'number', 'boolean'].includes(typeof item)) {
		output = item
	} else {
		try {
			if (Object.prototype.toString.call(item) === '[object Error]') {
				output = '\n' + item.stack.toString()
			} else {
				output = '\n' + JSON.stringify(item)
			}
		} catch (err) {
			console.log(err + '\n' + item)
			output = ''
		}
	}
	return output
}

function getDate() {
	const date = new Date()
	return {
		day: date.getFullYear() + '_' + date.getMonth() + '_' + date.getDate(),
		time: date.getHours() + ':' + date.getMinutes() + ':' + date.getSeconds(),
	}
}
