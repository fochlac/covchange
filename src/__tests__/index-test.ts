import logger from '../utils/logger'
import server from '../index'

jest.mock('../utils/logger')

describe('start server', () => {
	it('should start server', async () => {
		await server
		expect(logger).toBeCalledTimes(1)
		expect(logger).toMatchSnapshot()
	})
})
