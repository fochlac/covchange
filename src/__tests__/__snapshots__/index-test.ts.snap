// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start server should start server 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      0,
      "listening to http://localhost:51337/",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
